Transformations: Add Group to Nested Tables Transformation

codeincarnate opened the pull request:
**What is this feature?**

In previous versions, we added the ability within the table component to have subtables within rows. This is useful for a wide variety of use cases.  However, we've not had the ability to utilize this feature from the user interface _until now_.

This PR adds the Group to Subframe transformation. This transformation works similarly to the _Group by_ transformation, but rather than simply reducing values to a single row, it will leave ungrouped data fields as separate records which can be viewed as a subtable.


**Why do we need this feature?**

This is useful for a wide variety of use cases. For example, grouping a view requests by return status or HTTP verb. It can also be used for product analysis by category and type.


https://github.com/grafana/grafana/assets/199847/bf81dd18-1885-4d99-a16e-fcac66b5d787


Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

codeincarnate committed:
Stub group to subframe transformation

codeincarnate committed:
Get proper field grouping

codeincarnate committed:
Mostly working but fields not displaying 😭

codeincarnate committed:
Fix display processing in nested tables

codeincarnate committed:
Merge branch 'main' into codeincarnate/group-to-subtable

codeincarnate committed:
Modularize and start merging groupBy and groupToSubrame

codeincarnate committed:
Get this working

codeincarnate commented:
Did substantial modifications to share as much from the `Group by` transformation as possible. Working through linting/test breakages now.

codeincarnate committed:
Prettier

codeincarnate committed:
Typing things

codeincarnate committed:
More types

codeincarnate commented:
Still need to add some tests, but would be good if folks can start looking from a functionality perspective.

@catherineymgui what do you think would be a good icon for this?

codeincarnate commented:
One thing currently thinking about is the naming, it does group to an actual sub-frame, which in the case of the table panel is a sub-table. However that may not always be the case, and so happy for thoughts there 😄

codeincarnate committed:
Merge branch 'main' into codeincarnate/group-to-subtable

baldm0mma commented on packages/grafana-data/src/transformations/transformers/groupBy.ts:
```diff
@@ -209,4 +168,62 @@ const detectFieldType = (aggregation: string, sourceField: Field, targetField: F
+
+/**
+ *
+ * @param frame
```
Looks like there's also a `groupByFields` param to add here?

codeincarnate committed:
Add option for showing subframe table headers

codeincarnate committed:
Prettier

codeincarnate committed:
Get tests going

codeincarnate commented:
Cleaned up the code and added the option to show headers (or not) in the subframe. Also bringing over some tests from `groupBy` but not done yet and going to tackle those next.

catherineymgui commented:
@codeincarnate re: icons - idk how I feel about these but what are your thoughts?
[New subgroup illustrations.zip](https://github.com/grafana/grafana/files/14088696/New.subgroup.illustrations.zip)

codeincarnate committed:
Update tests

codeincarnate commented:
Thanks so much @catherineymgui these icons are great!

codeincarnate committed:
Fix naming and add icons

codeincarnate commented on packages/grafana-data/src/field/fieldOverrides.ts:
```diff
@@ -206,6 +206,13 @@ export function applyFieldOverrides(options: ApplyFieldOverrideOptions): DataFra
         for (const nestedFrames of field.values) {
           for (let nfIndex = 0; nfIndex < nestedFrames.length; nfIndex++) {
             for (const valueField of nestedFrames[nfIndex].fields) {
+              // Get display processor for nested fields
```
PR also fixes this bug in which display process is not applied to field in nested frames

codeincarnate committed:
Betterer fix

codeincarnate committed:
Prettier

codeincarnate committed:
Fix CSS object syntax

codeincarnate committed:
Prettier

baldm0mma commented on packages/grafana-data/src/transformations/transformers/groupToSubframe.test.ts:
```diff
@@ -0,0 +1,243 @@
+import { groupToSubframeTransformer, GroupToSubframeTransformerOptions } from './groupToSubframe';
+import { DataTransformerID } from './ids';
+
+describe('GroupToSubframe transformer', () => {
```
Tests. SO hot right now. Tests. ❤️

codeincarnate commented:
Don't have a test dashboard, actually trying to avoid that in this case to get a more real experience of how this would be used. It should be fairly easy to test, as long as there's one field that has duplicate values that can be grouped on, along with at least one other field that should work out. With greater than two fields it's possible to test aggregations as well.

nmarrs committed:
Merge remote-tracking branch 'origin' into codeincarnate/group-to-subtable

nmarrs reviewed (commented):
Overall this new transformation is looking really good and valuable!

I did notice that there seems to be a bug with the transformation returning 1 too many sub frames, which end up in the panel crashing when trying to open the undefined sub frame


https://github.com/grafana/grafana/assets/22381771/798e81da-eeb4-4038-ad1e-5983d798b0ae

I'll also include the test dashboard I created for this (mirroring your added test case 😬)
[sub table subbb-1706864030945.json](https://github.com/grafana/grafana/files/14137766/sub.table.subbb-1706864030945.json)


The only other note I have is need to add documentation - though that can be in a follow-up PR :) great work on this!

codeincarnate commented on packages/grafana-data/src/transformations/transformers/groupBy.ts:
```diff
@@ -75,64 +79,19 @@ export const groupByTransformer: DataTransformerInfo<GroupByTransformerOptions>
-            if (shouldGroupOnField(field, options)) {
-              groupByFields.push(field);
-            }
-          }
```
It's a tricky situation. On one hand, if we don't refactor we can't reuse the code. But if we do refactor we have the possibility of introducing bugs. I'll give it a thorough run through (and poke at @leeoniya to do the same since he requested the refactor 😆 )

torkelo commented:
Wonder if we should really call it subframe vs group to nested table? not sure we use the word frame that much in the UI.

codeincarnate commented on public/app/features/transformers/editors/GroupToSubrameTransformerEditor.tsx:
```diff
@@ -0,0 +1,164 @@
+          <Select options={options} value={config?.operation} placeholder="Ignored" onChange={onChange} isClearable />
+        </div>
+
+        {config?.operation === GroupByOperationID.aggregate && (
```
For calculations to be applied grouping needs to be applied as well :smile:. This is something that was from `Group by`. I think we should probably add more sophisticated logic to both, but that may be best left for a follow-up PR.

codeincarnate replied:
Hmm, I'm having trouble replicating the JS error 🤔. Did that happen elsewhere?

codeincarnate commented on packages/grafana-data/src/transformations/transformers/groupToSubframe.ts:
```diff
@@ -0,0 +1,235 @@
+export const groupToSubframeTransformer: DataTransformerInfo<GroupToSubframeTransformerOptions> = {
+  id: DataTransformerID.groupToSubframe,
+  name: 'Group to Subframe',
+  description: 'Group and then place in a sub-frame',
```
Good suggestion!

codeincarnate commented:
Thanks @torkelo. I'd tend to agree, was perhaps overthinking it though thought there may be a future circumstance where having nested frames could be used in other contexts.

codeincarnate committed:
Stub alert for calcs with grouping, start renaming

codeincarnate committed:
Merge branch 'main' into codeincarnate/group-to-subtable

codeincarnate committed:
Add logic to show warning message for calculations

codeincarnate committed:
Add calc warning

codeincarnate committed:
Renaming and feature flag

codeincarnate committed:
Rename images

codeincarnate commented:
Made various updates, for testing this is now behind a feature flag called `groupToNestedTableTransformation`.

codeincarnate committed:
Prettier

codeincarnate committed:
Fix tests

codeincarnate committed:
Update feature toggle

codeincarnate committed:
Fix error showing extra blank row

codeincarnate commented:
Did get a fix in as well for the extra non-functional row :smile:

baldm0mma commented on packages/grafana-data/src/transformations/transformers/groupToNestedTable.ts:
```diff
@@ -0,0 +1,237 @@
+ *
+ * @param valuesByGroupKey
+ *  A mapping of group keys to their respective grouped values.
+ * @param options
```
Love the docs!

baldm0mma commented on public/app/features/transformers/editors/GroupByTransformerEditor.tsx:
```diff
@@ -49,8 +49,28 @@ export const GroupByTransformerEditor = ({
+  }
+  const showCalcAlert = hasAggregation && !hasGrouping;
+
+  console.log(options);
```
Is this `console.log()` supposed to be here?

nmarrs replied:
ah good catch, i'll push up a commit with a bit of code cleanup 👍

nmarrs reviewed (approved):
Just re-tested and everything looks good! Great job on this @codeincarnate :shipit:

nmarrs committed:
minor code cleanup

codeincarnate commented:
Thanks for the code cleanup, and apologies for my messiness 😬
